### ISSUE TITLE
Finding matching: the hard→soft identity ladder (limiting false positives)

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#2585]
# Finding matching: the hard→soft identity ladder (limiting false positives)

Design spec for extending the shared Finding matcher beyond its exact hard tier.
The ordered tier originated in #2584 as a single exact identity key with
order-preserving LCS matching, exhaustively proved against `IlBodyDiff`.
Identity-set matching has since landed. This issue specifies the remaining soft
ladder and — the load-bearing part — **how to add it without importing the
false-positive problem that sinks naive diff tools.**

Refs: spine #2564; ordered hard tier #2584; Finding roadmap #2614; coordinate model PR #2624.

## Coordinate prerequisite

PR #2624 defines the coordinate boundary this ladder builds on:

- collection order, not `Finding<T>.Ordinal`, is matching authority;
- `Ordinal` is optional retained metadata for ordered producers;
- identity-set producers do not fabricate ordinals;
- producer coordinates and cross-stream provenance remain typed payloads rather
  than one opaque generic anchor.

Soft matching therefore extends `FindingKey` correspondence with structured
identity projections, typed deltas, and match-tier provenance. Deterministic
tie-breaking uses input order or a producer's stable sort; it does not turn an
ordinal or representation-local coordinate into identity.

## Problem

The matcher (`FindingMatcher`) establishes which old occurrence corresponds to which new one.
Rung 0 matches only on **exact identity** (`FindingKey.IdentityKey` equality). That catches
adds/removes/exact-moves but misses every *graded* correspondence the real system needs:

- **IL**: a constant changed `ldc.i4 3 → 4` (today: remove+add; want: one `Changed` with the value delta); a body that changed by one instruction (want: near-miss match, not two unrelated hunks).
- **C#**: a statement whose trivia/formatting changed; `x += 1` vs `x = x + 1`; a renamed local.
- **Metadata/API**: a renamed member (same signature); a signature that evolved (added a defaulted param); an **extension method promoted to an instance method** (`Ext.DoThing(this MyType, int)` → `MyType.DoThing(int)`); a member **relocated** to a different type.

All of these are "the same thing, changed" — but only if the matcher can match on something *looser*
than exact identity. Soft matching is exactly where diff tools earn their false positives (spurious
renames/moves), so the ladder must be introduced with hard soundness rules.

## Model: one structured identity, a ladder of keys

Do **not** hand the matcher N independent canonicalizers. Each producer computes **one structured
canonical identity** (a facet bag) per occurrence, and the matcher keys on **projections** of it,
ordered strict→loose:

```text
identity(member) = { receiver, name, params, return, kind, nullable, refness, arity, ... }

tier 0 (hard):   project(all facets)                    exact identity          score 100
tier 1 (soft):   drop {kind}                            extension ≡ instance    score  ~85
tier 2 (soft):   drop {kind, nullable, refness}         annotation-only         score  ~80
tier 3 (soft):   project {receiver, name}               signature evolved       score  ~70
tier k (fuzzy):  similarity(body-shingle)/edit(name)    renamed / near body     score  ~50-65
```

- **More facets retained ⇒ harder ⇒ higher score ⇒ committed earlier.**
- The **domain knowledge lives in the one canonicalization** (the mechanical reprojection — e.g.
  stripping an extension method's `this` receiver, folding IL `ldc` encodings). The hard/soft ladder
  is just *which subset you key on*, so identity stays single-sourced (the "one implementation" goal).
- The softest rungs aren't facet subsets but **similarity measures** (locality-sensitive/shingle hash
  of an operation sequence; edit distance on names). So a rung is "a key **or** a similarity function."

This generalizes today's `FindingKey.IdentityKey: string` to an **identity ladder**
(hard key + ordered soft keys/measures). Rung 0 is the degenerate single-key case.

## Matching algorithm: committed-core-first, strict→loose

1. **Commit the hard tier.** Exact-identity mutual matches (LCS for ordered streams; order-free
   bijection for identity-set streams). These are removed from the residual first — this is the
   pilot's committed core, unchanged.
2. **Descend the ladder over the residual.** For each softer tier, generate candidate pairs among
   still-unmatched occurrences, scored by tier (harder = higher) plus corroboration (scope, neighbors).
   Each tier commits its mutual, unambiguous matches before the next looser tier runs.
3. **Resolve contention as a matching, not a filter.** Accepting old A ↔ new B forbids A ↔ C, so the
   fringe is a scored **bipartite matching**, resolved greedily by score with deterministic tiebreaks
   (this is the pilot's `FindingFold.ApplyAcceptance`, generalized across tiers).

The result stays a `FindingMatch` (`Matched`/`Moved`/`Added`/`Removed` links + a scored
fringe); soft matches surface as `Changed` (matched-but-different) or `Moved` rows carrying the tier
and the facet/similarity delta.

## The false-positive discipline (the point of this issue)

Soft matching must never make the diff *worse* than rung 0. Non-negotiable rules:

1. **Under-match is the default.** At the default acceptance threshold (100) **only the hard tier is
   committed** — soft matches are scored candidates that are *not* accepted unless a consumer opts in.
   So the out-of-the-box behavior is exactly the provable rung-0 behavior; soft matching is additive.
2. **A false equivalence is worse than a missed match.** A spurious soft match fabricates a `Changed`
   over two unrelated things and can hide a real add/remove; a missed match only loses a nicer label
   and degrades safely to add/remove. When unsure, **emit the weaker tier or no match.**
3. **Acceptance is a consumer fold, per dimension.** The threshold (and which tiers/similarity floors
   are allowed) is chosen by the consumer, not baked into the producer. A compat consumer may accept
   no soft matches; an exploratory consumer may accept renames flagged low-confidence, then verify.
4. **Determinism through ties.** Same inputs ⇒ same correspondence, including
   tie resolution from stable input order or a producer's stable sort. A
   non-deterministic matcher is not gateable.
5. **Near-miss NEGATIVE fixtures are the primary test surface.** The pilot's "mismatch resistance"
   gate generalizes to "soft-match resistance": two occurrences a naive soft scorer would fuse must
   stay split at the default. Every soft tier ships with negatives (things that look similar but must
   not match) *before* positives.
6. **Soft never overrides hard.** Hard-tier commits happen first and remove their occurrences from the
   residual, so a soft candidate can never displace an exact match.

## Per-stream instantiation

| Stream | Hard key (tier 0) | Representative soft tiers |
| --- | --- | --- |
| **IL** | opcode-family + operand identity (== `CanonicalEquals`) | drop immediate value (`ldc 3→4` ⇒ `Changed`); shingle-hash of the op sequence (near-identical body) |
| **C#** | canonical statement shape (trivia folded) | trivia/whitespace (`EncodingOnly`); sugar/paren (`x+=1` vs `x=x+1`); local rename; statement-shingle |
| **Metadata/API** | canonical signature (declaring type + kind + params + return) | drop `kind` (extension→instance); drop nullable/refness; `{receiver,name}` (signature evolved); name edit-distance (rename); drop declaring-type (relocation) |

Note the **ordered vs identity-set** split: IL and C# are ordered (enumeration
order matters; moves are relocations in a sequence); Metadata/API is a set (no
order; a future soft "move" is a `ScopeKey`/declaring-type change). Both ride
the same matcher via `FindingMatchMode`.

## Scaling & memory (foundational, from #2584 round-5 review)

The pilot's ordered committer is a full-matrix LCS: `O(N·M)` time **and space**. That is fine at
method-body scale (the shipped `IlFindings` only ever feeds single bodies), but it does not scale to
assembly-scale streams — an unordered 50k-member Metadata stream would try to allocate a ~10 GB
matrix. The matcher now:

- route **large unordered** streams through the **identity-set committer** (hash the hard keys →
  `O(N)` bijection; no LCS at all) — this is the primary answer for Metadata/API scale, and
- still needs a **linear-space ordered** algorithm (Hirschberg, or Myers `O(ND)` which also scales when the
  edit distance is small) for large **ordered** streams, replacing the full matrix.

Until then, the ordered matcher **guards** its matrix with a documented cell cap and fails with a
clear error rather than OOM (shipped in #2584). Move-run and fringe passes are `O(R²)` over the
residual `R`; acceptable at body scale (~2.5 s at R=15k, deterministic), but the identity-set path
avoids them for large streams.

## Ordered vs identity-set is a first-class mode, not an afterthought

`FindingMatchMode` is already first-class. `Ordered` runs LCS and enforces
original-stream contiguity in the move pass. `IdentitySet` bypasses LCS and
contiguity entirely and commits a hard-key bijection, so a permutation is a
non-event. Soft tiers must preserve this split: sequence relocation belongs to
ordered matching, while a Metadata relocation is a typed identity/scope delta.
Both modes should share the same tier ladder and fringe/acceptance machinery.

## Leaf API implications

- `FindingKey`: `IdentityKey: string` → an **identity ladder** (hard key + ordered soft
  keys and/or similarity descriptors). Keep a one-key convenience for rung-0 producers.
- `FindingMatcher`: tiered commit loop (hard → soft → fuzzy); today's LCS becomes the tier-0
  ordered committer, with an order-free committer for identity-set streams.
- `FindingMatchOptions`: add enabled tiers, per-tier score floors, and similarity thresholds;
  retain the existing `FindingMatchMode`.
- Acceptance stays `FindingFold`/consumer-side; the fringe carries the tier + delta so a consumer
  can decide and a presenter can explain ("renamed", "signature evolved", "moved Ext→MyType").
- **Value-object hygiene:** aggregate/result types that carry `ImmutableArray`/`ImmutableHashSet`
  (`FindingMatch`, `FindingEquivalence`, `FindingComparison<CanonicalIlOperation>`) currently get record-generated
  *reference* equality over those collections, not structural equality — so two logically-identical
  results compare unequal. Before consumers treat these as value objects (dedup, cache keys, "did the
  diff change?"), give them explicit `SequenceEqual`/`SetEquals` equality or stop advertising value
  semantics. Decide per type during the rewrite.

## Relationship to the three identity axes

The same ladder serves cross-**version** (the matcher here), cross-**producer** (`FindingSubject.Key`
agreement), and cross-**stream** (IL↔C# provenance) — but the *evidence source* differs and must stay
honest: cross-stream is authoritative-by-provenance (no scoring), cross-producer is exact-checked (a
mismatch is a bug, census-ratcheted), and only cross-**version** is genuinely fuzzy and should climb
the soft ladder. Don't score what you already know.

## Phasing (rungs)

- **Rung 0 (done, #2584):** exact identity, ordered LCS, under-claim default, acceptance-as-fold.
- **Rung 1 (done):** identity-set (order-free) committer for Metadata/API; still exact tier only.
- **Rung 2:** facet-subset soft tiers (drop kind/nullable/etc.) with negative fixtures; extension→instance and signature-evolution as `Changed`.
- **Rung 3:** fuzzy similarity tiers (name edit-distance, body shingle) with strict floors + heavy negatives.
- Each rung: reproduce the prior rung's output when its new tiers are disabled (superset property), and ship negatives before positives.

## Open questions

- Similarity representation: shingle/w-shingling vs minhash vs edit-distance — per stream, and where the hash lives (producer canonicalizer).
- Should tier scores be a fixed ladder or learned/tunable per corpus? (Start fixed + documented.)
- Hierarchical occurrences (API type-level *and* member-level; C# statement *and* expression) — do soft tiers cross levels?
- Global-optimal vs greedy matching for the fringe — greedy+deterministic is proposed; is any stream worth a Hungarian-style optimum?

## Acceptance criteria

- A documented, deterministic tier ladder with per-tier scores and a default that equals rung-0 behavior.
- Negative ("soft-match resistance") fixtures for every soft tier, added before the positives.
- Superset property: with soft tiers disabled, output equals the exact-only matcher.
- Extension→instance, signature-evolution, and IL `ldc 3→4` render as single `Changed`/`Moved` rows under an opt-in threshold, and as add/remove at the default.

**Comment by richlander** [comment]
## Representation and discriminating power (clarified) + two worked scenarios

Sharpening the "one structured identity + projections" line, because the representation choice and the false-positive discipline turn out to be the same decision.

### Representation: one canonicalization; the matcher matches on *less* per rung

Not a precomputed vector of keys on the occurrence. The occurrence carries **one structured identity** — a facet bag (`name -> comparable value`) plus raw similarity features (name text, body shingle/minhash) — and the **producer supplies the rung ladder as data** (which facets a rung keys on, or which similarity fn/feature/floor). The matcher is domain-free: per rung it *projects* the named facet subset to a key (exact rungs) or applies the named similarity fn (fuzzy rungs), committing hardest-first.

A vector-of-keys is only a **derived cache** the matcher may materialize for the exact rungs — it can't be the primary model because (a) fuzzy rungs aren't equality keys (they need the raw feature / an LSH signature), and (b) opaque keys can't *explain* a soft match's delta, which a false-positive-averse system needs for auditing and rendering. Today's scalar `IdentityKey` is exactly `hash(project(all facets))` — the tier-0 key of a one-rung ladder.

### Discriminating power has four loci — intrinsic identity alone is provably insufficient

1. **Intrinsic facets** (content/signature/opcode) — collide exactly where it matters (two identical helpers/instructions).
2. **Relational context** (declaring scope, enclosing region, neighbor fingerprints, position) — breaks intrinsic collisions. The pilot's `ScopeKey` + run-contiguity is a primitive version.
3. **Match-time constraints** (bijection, mutual-best, global order-consistency) — live in the matcher.
4. **Attach re-validation** — recover facets *deliberately dropped* for match robustness (the pilot's branch-target check is exactly this: IL identity folds the branch target, Attach re-checks it).

Two principles make the ladder a filter *by construction*:

- **Tier 0 is a complete fingerprint** (keys on everything). It only fuses truly-identical occurrences, so all false-positive risk is confined to softer rungs where you *explicitly* drop a named facet; Attach recovers anything a drop hid.
- **Rung order is mostly forced by the facet-subset lattice** (supersets commit first); measured per-rung precision on a near-miss corpus breaks ties between incomparable rungs, and unvalidated rungs ship disabled.

The reliability criterion, distilled: **a match survives a set of transformations iff at least one strong anchor survives them.** The two scenarios below are the same rule with the anchor present vs absent.

### Scenario A — extension method promoted to an instance method (anchor survives)

`Ext.DoThing(this MyType x, int y)` → `MyType.DoThing(int y)`. The receiver `MyType` lives in *different metadata slots* (first param vs declaring type), so a naive `declaring-type + all-params` fingerprint has **no shared facet** and no projection can align them. The canonicalizer must **lift the receiver into its own `receiver` facet** (from the `this` param for extensions, from the declaring type for instance methods), same slot for both:

```
extension: { receiver: MyType, declaring: Ext,    name: DoThing, params: [int], return: bool, kind: Extension }
instance:  { receiver: MyType, declaring: MyType,  name: DoThing, params: [int], return: bool, kind: Instance  }
```

- **Tier 0** keeps `receiver` + `declaring` + `kind` → the two are distinct → no false fusion. The reprojection is *lossless* (receiver relocated, not dropped), so completeness holds.
- **Soft rung** drops `{declaring, kind}`, **keeps `receiver`** → they match on the shared anchor → reported as **move** (`declaring: Ext→MyType`) **+ re-kind** (`Extension→Instance`).
- **Reliability** = the receiver is the semantically meaningful shared anchor; matching on `{receiver,name,params,return}` is strong. Anchor *specificity* governs confidence: `this MyType` strong; `this object`/`this IEnumerable<T>` weak → lower score, lean under-claim.

The canonicalization choice (lift the receiver) and the soft rung's reliability are the same decision.

### Scenario B — a constant that both moves scope and changes value (anchor must come from outside)

`ldc.i4 4` in method scope → `ldc.i4 2` inside an `if` block. Both of the constant's own identity axes change — **content** (`4→2`) and **context** (`method→if`). The only surviving intrinsic facet is the opcode family `ldc.i4`, shared by every constant load — far too weak. So the constant **cannot be matched on its own merits**; standalone, the correct result is under-claim → `Removed ldc.i4 4` + `Added ldc.i4 2`.

It *is* matchable when a stable anchor survives *outside* the constant — its def-use / neighbor:

```
old (method):   ... ; ldc.i4 4 ; stloc.0 ; ...
new (if-block): ... ; brfalse ... ; ldc.i4 2 ; stloc.0 ; ...
```

`stloc.0` is identical on both sides. A rung keying on `{opcode-family, right-neighbor-fingerprint}` and dropping `{value, scope}` matches the constant *via its consumer*, and reports **both** transformations as deltas of the one matched occurrence: value `4→2` **and** scope `method→if`.

Really the right unit is the **enclosing assignment**: `x = 4` (method) ↔ `x = 2` (in the `if`), matched on the target `x` (strong anchor), with the constant change as a value-delta *inside* the matched statement and the relocation as the statement's scope-move — the **hierarchical-occurrence** point above. The leaf constant is matched *because its parent matched*.

**The limit is "≥1 strong anchor survives," not "≤1 transformation."** Scenario A survives because the receiver holds; Scenario B survives only if an external use/neighbor holds, and not at all if the constant is bare — where under-claim is the correct, honest answer, and the whole value of the match (when it fires) is collapsing two orphan add/remove rows into one legible finding: "the assignment to `x` moved into the `if` and its value went 4→2."

Refs #2564, #2584.

**Comment by richlander** [comment]
## Review: three Rung-2 blockers + three clarifications

The spec plus the representation/discriminating-power comment are close to implementable — the **"a match survives iff ≥1 strong anchor survives"** criterion, the four loci, the receiver-lift, and Attach-re-validation are the right frame. Skipping what's covered; here are the genuinely-open edges, weighted. The first three I'd treat as **blocking for Rung 2**, because each is a soundness-or-scale property that's cheap to design in now and expensive to retrofit once soft tiers ship.

### 1. Match confidence must survive onto the *row*, not just the link

In the pilot, `EvidenceLink` carries `Confidence` (100 committed, lower for accepted fringe) — but `EvidenceFold.ToRow` **drops it**; the projected `EvidenceRow` has no confidence/tier field. Once soft matching lands, `Changed` does double duty: a **hard** `Changed` (exact identity, body differs — e.g. branch retarget) and a **tier-3 fuzzy** `Changed` (probably-same, opt-in, low confidence) render identically. The fringe carries tier+delta, but the **accepted row must too**, or a consumer/presenter can't distinguish "definitely the same, signature changed" from "guessed, 60% confident." Add `Confidence` (+ the facet/similarity delta, and which tier accepted it) to the row model.

### 2. Soft/fuzzy tiers need *blocking* — the scaling section only guards the hard matrix

The scaling fixes (identity-set bijection, Hirschberg) cover tier 0. But soft tiers generate candidates **all-pairs, `O(tiers · R²)`** over the residual, and fuzzy similarity adds per-pair cost. On a large residual with soft tiers enabled that blows up *independent* of the hard-tier fix. The open question names minhash — commit it as **LSH blocking** (bucket by minhash band, compare only within buckets), not merely as a similarity representation, and give the soft passes the same cell-cap guard the hard matrix now has (#2584 findings 7/10). The opt-in default hides this, but a consumer enabling renames on a rewritten assembly hits it.

### 3. Relational/neighbor anchors must draw only from *already-committed* matches

Scenario B leans on `stloc.0` as the surviving anchor — which works precisely because it's hard-matched. But "neighbor-fingerprint" corroboration is circular if it can key on *other soft candidates* (A anchors on B, B on A → non-deterministic, order-dependent). State the rule: **relational/neighbor corroboration anchors only on matches committed by a harder tier, never same-or-looser candidates.** That's what keeps the soft ladder acyclic and gateable — patience/histogram diff use exactly this (unique committed anchors first).

### Clarifications (worth stating / answering open questions)

**4. Separate the two axes "score" conflates.** Commit *order* is the facet-subset **lattice** (supersets first) with per-rung measured precision breaking incomparable ties — the comment nails this. But the numeric scores (85/80/70) are a *consumer-facing confidence*, a different axis. "Harder = higher score = committed earlier" reads as one number doing both, yet a scope-corroborated soft match can have high confidence while still committing *after* a harder tier. Make it explicit: **ordering = tier lattice; the number = confidence for the acceptance fold.**

**5. Open question answered — greedy is right, and here's why.** Because the soft layer is *advisory and under-claimed*, a suboptimal greedy pairing degrades to add/remove — the safe direction — so Hungarian optimality isn't worth the non-locality and complexity. Keep greedy + deterministic; reserve an optimal matcher only if a specific stream's negative fixtures show greedy mis-pairing. The under-claim discipline is what licenses greedy.

**6. C# is both ordered *and* identity-set, per level.** The per-stream table labels C# "ordered," but statement streams are ordered while a type's *member list* is set-like — same split as Metadata. So `EvidenceStreamKind` is chosen per match-invocation / occurrence-level, not per language; C# spans both. Ties into the hierarchical-occurrence open question.

### Net

Rungs and discipline are right; #1–3 are the load-bearing gaps I'd close before Rung 2 ships soft tiers. #1 is a small row-model change, #2 is a scaling landmine hidden by the opt-in default, #3 is the acyclicity rule that makes neighbor-corroboration gateable.

Refs #2564, #2577, #2584.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by richlander** [comment]
## Accepted — folding the three blockers + three clarifications into the spec

All six are right. Dispositions:

**1. Confidence must survive onto the row — accepted, and it's implementable now (not speculative).** `EvidenceMatchEntry.Confidence` already exists; `EvidenceFold.ToRow` drops it, so even today a fringe-accepted `Moved` row (lowered threshold) loses its 50/75 confidence. Propagating the existing `Confidence` onto `EvidenceRow` is fixing an information drop, not adding a speculative field — so it lands in the **matcher-foundation PR now**. The *tier* and *facet/similarity delta* fields are deferred to when soft tiers exist (those would be speculative today), but `Confidence` goes on the row now, and a hard `Changed` (confidence 100) vs a future fuzzy `Changed` (e.g. 60) will then be distinguishable.

**2. Soft/fuzzy tiers need blocking, not just a similarity representation — accepted as a Rung-2 requirement.** Minhash is promoted from "representation" to **LSH blocking**: bucket by minhash band, compare only within buckets; and every soft pass inherits the same cell-cap guard the hard matrix got in #2584 (findings 7/10). The all-pairs `O(tiers·R²)` residual cost is a real landmine the opt-in default hides.

**3. Neighbor/relational corroboration anchors only on harder-committed matches — accepted as the acyclicity rule.** Stated: relational corroboration may key **only** on matches already committed by a strictly harder tier, never on same-or-looser candidates. That's what keeps the ladder acyclic, order-independent, and gateable (patience/histogram diff's "unique committed anchors first"). Note the pilot is already compliant — its fringe corroborates on `ScopeKey`, not on other candidates.

**4. Two axes, not one — accepted; correcting the spec.** Commit **order** = the facet-subset lattice (supersets first) + measured precision for incomparable ties. The numeric **score** = consumer-facing confidence for the acceptance fold. They're independent: a scope-corroborated soft match can be high-confidence yet still commit after a harder tier. The earlier "harder = higher score = committed earlier" wording conflated them and is retracted.

**5. Greedy is the answer (open question closed).** Because the soft layer is advisory and under-claimed, a suboptimal greedy pairing degrades to add/remove — the safe direction — so Hungarian optimality isn't worth the non-locality. Keep greedy + deterministic; only reach for an optimal matcher if a specific stream's negative fixtures show greedy mis-pairing.

**6. StreamKind is per-level, not per-language — accepted; correcting the per-stream table.** C# statement streams are ordered; a type's member list is set-like — same split as Metadata. So `EvidenceStreamKind {Ordered, IdentitySet}` is a **per-match-invocation / per-occurrence-level** choice, not a fixed per-producer attribute. This is why `StreamKind` belongs on the match options, and it ties into the hierarchical-occurrence question (statement-level ordered vs member-level set within one C# producer).

### Sequencing impact

- **Matcher-foundation PR (now, mine):** `EvidenceStreamKind` + the identity-set committer (per-invocation, per #6); `Confidence` propagated onto `EvidenceRow` (#1); the acyclic-corroboration rule (#3) and ordering-vs-score split (#4) encoded as the matcher's contract even before soft tiers.
- **Rung 2 (later):** facet-subset soft tiers + LSH blocking (#2) + tier/delta on the row, with negative fixtures per tier.

Thanks — #1–3 would have been painful retrofits. Refs #2564, #2584.

**Comment by richlander** [comment]
## Registry design spike — capability + lazy consumer-strategy (runnable)

Following the design thread on the `Finding → Fact → Triage` ladder and the `Finding<T>` payload triad (#2591), here is a **runnable prototype** of the registry that lets the aggregation layer (Research) reach payloads it doesn't statically know the type of.

**Locked shape** (synthesised from the two in-house references):
- **Bridge (from markout `MarkoutTypeInfo<T>`/`IMarkoutTypeInfo` + smt `prototype-registry.cs` V3):** a non-generic capability interface `IFactFolder` (the aggregation view) over a generic base `FactFolder<T>` where the `(Finding<T>)row` cast is localized and invariant-safe.
- **Capabilities, not one uniform op (from smooth-markdown-table):** a shape *opts into* what it can do. Only shapes that can fold to a Fact register a folder; `il.op` cannot and simply contributes no facts. markout's single uniform `Serialize` cannot express "this type does not support this transform."
- **Lazy, consumer-selected strategy (from smt strategy APIs):** asking for a skeleton summary runs no fold; asking for the Fact tier folds **only** capable shapes, **only** then.
- Registry keyed by `Descriptor.Id` (what Research reads off the skeleton), recording **tier**. Hand-listed now; graduates to a source-gen context (markout `MarkoutSerializerContext`) with the *same* shapes — no rework.

### Output
```
== Strategy A: skeleton summary (no payload, no fold) ==
  5 findings across 2 descriptors

== Strategy B: Fact tier (lazy fold; only capable shapes) ==
    [fold ran] projecting 3 raw alloc findings -> facts
  [Fact] semantic.allocation: A.Run — 2 allocs, some in loop
  [Fact] semantic.allocation: A.Init — 1 allocs
  (il.op CanFold=False -> contributed 0 facts)

== Same-domain typed access (no registry, no cast) ==
  typed payload: A.Run box inLoop=True
```

Note the `[fold ran]` probe fires only under Strategy B (lazy), and `il.op` (no capability) yields 0 facts.

### Prototype
```csharp
// finding-registry.cs — DESIGN SPIKE (throwaway, file-based app) for issue #2585.
//
// Pressure-tests the Finding registry shape, combining the two in-house references:
//   * markout .............. the non-generic-iface + generic-base (T)row BRIDGE, and the
//                            source-gen context graduation path (MarkoutSerializerContext).
//   * smooth-markdown-table  CAPABILITY interfaces (a shape opts into what it can do) and
//                            CONSUMER-strategy-driven, LAZY consumption.
//
// It demonstrates, end to end:
//   (A2) T-recovery bridge: IFactFolder (non-generic, aggregation view) + FactFolder<T>
//        (generic base) where the (Finding<T>)row cast is localized, invariant-safe.
//   (C1) metadata self-described per shape (here: instance props; see graduation note).
//   (D2) transforms as CAPABILITIES: only shapes that CAN fold to a Fact register a folder;
//        IL-op findings cannot, and simply contribute no facts (markout's single uniform
//        Serialize cannot express "this type does not support this transform").
//   (E2) LAZY, consumer-selected strategy: asking for a skeleton summary runs no fold; asking
//        for the Fact tier runs folds ONLY for capable shapes, and only then.
//   Registry keyed by Descriptor.Id (what Research can read off the skeleton), recording TIER.
//
// Run: dotnet run finding-registry.cs

using System.Collections.Immutable;

// ---------------------------------------------------------------------------
// 1. Substrate (mirrors src/ILInspector.Findings, trimmed) — the Finding<T> triad from #2591.
// ---------------------------------------------------------------------------

enum FindingTier { Finding, Fact, Triage }
enum FindingKind { Present, Added, Removed, Changed }

record FindingSubject(string Key, string Display);
record FindingDescriptor(string Id, string Title);
record FindingAnchor(string IdentityKey, int OldIndex = -1, int NewIndex = -1);

/// <summary>Domain-free skeleton — the aggregation currency. Research holds this.</summary>
interface IFinding
{
    FindingSubject Subject { get; }
    FindingDescriptor Descriptor { get; }
    FindingKind Kind { get; }
    FindingAnchor Anchor { get; }
    string? Detail { get; }
}

/// <summary>Typed payload for same-domain consumers (no registry, no cast).</summary>
interface IFinding<T> : IFinding
{
    T? Payload { get; }
}

sealed record Finding<T>(
    FindingSubject Subject,
    FindingDescriptor Descriptor,
    FindingKind Kind,
    FindingAnchor Anchor,
    string? Detail = null,
    T? Payload = default) : IFinding<T>;

// ---------------------------------------------------------------------------
// 2. Domain payloads (mirror real producers, trimmed).
// ---------------------------------------------------------------------------

// Raw, TERMINAL: an IL-op finding does not fold to a Fact.
record IlOp(string Opcode, string? Operand);

// Raw, but FOLDS to an AllocationFact (N raw findings -> 1 fact).
record AllocationOccurrence(string Method, string Kind, bool InLoop);

// The Fact produced by folding allocation findings for a method.
record AllocationFact(string Method, int Count, bool AnyInLoop);

// ---------------------------------------------------------------------------
// 3. Capability: "folds to a Fact" (D2). A shape opts in by registering a folder.
// ---------------------------------------------------------------------------

/// <summary>Non-generic capability — the aggregation view. Research calls this holding IFinding.</summary>
interface IFactFolder
{
    string SourceDescriptorId { get; }             // which raw finding shape it consumes
    FindingDescriptor ProducedDescriptor { get; }  // the Fact it emits (registered at tier = Fact)
    IReadOnlyList<IFinding> Fold(IReadOnlyList<IFinding> raw);
}

/// <summary>
/// Generic base — the (Finding&lt;T&gt;)row cast lives HERE, authored where T is known and
/// guaranteed safe by the descriptor-&gt;T invariant. This is the bridge (markout / smt V3).
/// </summary>
abstract class FactFolder<T> : IFactFolder
{
    public abstract string SourceDescriptorId { get; }
    public abstract FindingDescriptor ProducedDescriptor { get; }

    public IReadOnlyList<IFinding> Fold(IReadOnlyList<IFinding> raw)
    {
        var typed = new List<Finding<T>>(raw.Count);
        foreach (var row in raw)
            typed.Add((Finding<T>)row);            // localized, invariant-safe downcast
        return FoldTyped(typed);
    }

    protected abstract IReadOnlyList<IFinding> FoldTyped(IReadOnlyList<Finding<T>> raw);
}

// ---------------------------------------------------------------------------
// 4. A concrete capability implementation (producer-authored, knows T).
// ---------------------------------------------------------------------------

sealed class AllocationFactFolder : FactFolder<AllocationOccurrence>
{
    public override string SourceDescriptorId => "alloc";
    public override FindingDescriptor ProducedDescriptor => new("semantic.allocation", "Allocation semantic fact");

    protected override IReadOnlyList<IFinding> FoldTyped(IReadOnlyList<Finding<AllocationOccurrence>> raw)
    {
        Console.WriteLine($"    [fold ran] projecting {raw.Count} raw alloc findings -> facts");  // laziness probe
        return raw
            .Where(r => r.Payload is not null)
            .GroupBy(r => r.Payload!.Method)
            .Select(g => (IFinding)new Finding<AllocationFact>(   // typed .Payload read: no box, no cast
                new FindingSubject(g.Key, g.Key),
                ProducedDescriptor,
                FindingKind.Present,
                new FindingAnchor(g.Key),
                Detail: $"{g.Count()} allocs{(g.Any(x => x.Payload!.InLoop) ? ", some in loop" : "")}",
                Payload: new AllocationFact(g.Key, g.Count(), g.Any(x => x.Payload!.InLoop))))
            .ToList();
    }
}

// ---------------------------------------------------------------------------
// 5. Registry — hand-listed, keyed by Descriptor.Id. Records TIER and the fold CAPABILITY.
//    Graduation note: replace the ctor lists with a source-gen context (markout
//    MarkoutSerializerContext) reading static-abstract shape metadata; SAME shapes, no rework.
// ---------------------------------------------------------------------------

sealed class FindingRegistry
{
    readonly Dictionary<string, FindingTier> _tiers;
    readonly Dictionary<string, IFactFolder> _folders;

    public FindingRegistry(
        IReadOnlyList<(string DescriptorId, FindingTier Tier)> tiers,
        IReadOnlyList<IFactFolder> folders)
    {
        _tiers = tiers.ToDictionary(t => t.DescriptorId, t => t.Tier);
        _folders = folders.ToDictionary(f => f.SourceDescriptorId);
        foreach (var f in folders)
            _tiers[f.ProducedDescriptor.Id] = FindingTier.Fact;   // a folder's output is Fact-tier
    }

    public FindingTier TierOf(string descriptorId) => _tiers.GetValueOrDefault(descriptorId, FindingTier.Finding);
    public bool CanFold(string descriptorId) => _folders.ContainsKey(descriptorId);   // capability query
    public IFactFolder? FolderFor(string descriptorId) => _folders.GetValueOrDefault(descriptorId);
}

// ---------------------------------------------------------------------------
// 6. Research aggregation — CONSUMER picks the strategy (tier); work is LAZY (E2).
// ---------------------------------------------------------------------------

static class Research
{
    // Strategy A: skeleton-only summary. No payload touched, no fold. (smt "Skip"/"Metrics")
    public static string Summarize(IReadOnlyList<IFinding> findings)
        => $"{findings.Count} findings across {findings.Select(f => f.Descriptor.Id).Distinct().Count()} descriptors";

    // Strategy B: elevate to the Fact tier. Folds only CAPABLE shapes, and only when asked.
    public static IReadOnlyList<IFinding> ToFacts(IReadOnlyList<IFinding> findings, FindingRegistry registry)
    {
        var facts = new List<IFinding>();
        foreach (var group in findings.GroupBy(f => f.Descriptor.Id))
        {
            var folder = registry.FolderFor(group.Key);
            if (folder is null)
                continue;   // no fold capability (e.g. il.op) -> contributes no facts
            facts.AddRange(folder.Fold([.. group]));
        }
        return facts;
    }
}

// ---------------------------------------------------------------------------
// 7. Demo
// ---------------------------------------------------------------------------

class Program
{
    static void Main()
    {
        // Two producers emit RAW findings (Finding tier), heterogeneous payloads.
        IFinding[] il =
        [
            new Finding<IlOp>(new("M", "M"), new("il.op", "IL operation"), FindingKind.Present, new("ldarg.0", 0, 0), Payload: new IlOp("ldarg.0", null)),
            new Finding<IlOp>(new("M", "M"), new("il.op", "IL operation"), FindingKind.Added, new("call", -1, 3), Payload: new IlOp("call", "Foo")),
        ];
        IFinding[] alloc =
        [
            new Finding<AllocationOccurrence>(new("A.Run", "A.Run"), new("alloc", "Allocation"), FindingKind.Present, new("box#1"), Payload: new("A.Run", "box", InLoop: true)),
            new Finding<AllocationOccurrence>(new("A.Run", "A.Run"), new("alloc", "Allocation"), FindingKind.Present, new("arr#1"), Payload: new("A.Run", "array", InLoop: false)),
            new Finding<AllocationOccurrence>(new("A.Init", "A.Init"), new("alloc", "Allocation"), FindingKind.Present, new("box#2"), Payload: new("A.Init", "box", InLoop: false)),
        ];
        IReadOnlyList<IFinding> all = [.. il, .. alloc];

        var registry = new FindingRegistry(
            tiers: [("il.op", FindingTier.Finding), ("alloc", FindingTier.Finding)],
            folders: [new AllocationFactFolder()]);

        Console.WriteLine("== Strategy A: skeleton summary (no payload, no fold) ==");
        Console.WriteLine("  " + Research.Summarize(all));

        Console.WriteLine();
        Console.WriteLine("== Strategy B: Fact tier (lazy fold; only capable shapes) ==");
        var facts = Research.ToFacts(all, registry);
        foreach (var f in facts)
            Console.WriteLine($"  [{registry.TierOf(f.Descriptor.Id)}] {f.Descriptor.Id}: {f.Subject.Display} — {f.Detail}");
        Console.WriteLine($"  (il.op CanFold={registry.CanFold("il.op")} -> contributed 0 facts)");

        Console.WriteLine();
        Console.WriteLine("== Same-domain typed access (no registry, no cast) ==");
        var one = (IFinding<AllocationOccurrence>)alloc[0];
        Console.WriteLine($"  typed payload: {one.Payload!.Method} {one.Payload.Kind} inLoop={one.Payload.InLoop}");
    }
}
```

### Side note (not part of this work)
Building this surfaced a **performance opportunity in markout**: its context folds eagerly and uniformly. The capability + lazy-strategy shape here could let markout skip projections a consumer never requests. Tracked separately as a markout idea — no action here.

**Comment by richlander** [comment]
Captured the user-requested three-population History scenario in #7454: instance T.M -> extension TExtensions.M -> the same extension declaring Type forwarded to another Library within the Package. This adds a temporal consumer scenario for the declaration-form transition already contemplated here, not approval to weaken strict Metadata correspondence. The History baseline and typed migration-evidence questions remain explicit in #7454.
